### PR TITLE
Extend MO for the operation NonMaxSuppression-5

### PR DESCRIPTION
**Pull request by vgavrilo**
Extend MO for the operation NonMaxSuppression-5.
Issue: #37658

### CODE REVIEW

File: model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py [hunk 1]
```diff
@@ -30,10 +30,10 @@ class NonMaxSuppressionNormalize(FrontReplacementSubgraph):
 
     def find_and_replace_pattern(self, graph: Graph):
         for nms in graph.get_op_nodes(op='NonMaxSuppression'):
-            # make inputs 2 to 4 to have shape [] instead of [1] (convert 1D to 0D)
+            # make inputs 2 to 4 to have shape [1] instead of [0] (convert 0D to 1D)
+            nms_name = nms.soft_get('name', nms.id)
             for port_id in range(2, 5):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py:35]: 5 must be changed to 6 since inputs with indices 2, 3, 4 and 5 must be updated

File: model-optimizer/extensions/ops/non_max_suppression.py [hunk 2]
```diff
@@ -57,31 +56,53 @@ def backend_attrs(self):
 
     @staticmethod
     def infer(node: Node):
+        num_of_inputs = len(node.in_ports())
+        assert 2 <= num_of_inputs <= 6, \
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/non_max_suppression.py:60]: 6 inputs can be in the opset5 version of the operation only. Other version of the operation must have less.

File: model-optimizer/extensions/ops/non_max_suppression.py [hunk 3]
```diff
@@ -57,31 +56,53 @@ def backend_attrs(self):
...
+            node.out_port(2).data.set_shape(np.array(1).astype(np.int64))
+
     @staticmethod
     def type_infer(node):
-        if node.get_opset() in ['opset3', 'opset4']:
+        opset = node.get_opset()
+        if opset == 'opset5':
+            node.out_port(0).set_data_type(node.output_type)
+            num_of_outputs = len(node.out_ports())
+            if num_of_outputs >= 2:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/non_max_suppression.py:101]: Check that output port #1 is connected instead. 

Similar comment for the line 103.

File: model-optimizer/extensions/front/tf/non_max_suppression_normalize.py [hunk 4]
```diff
@@ -64,18 +64,34 @@ def replace_sub_graph(graph: Graph, match: dict, **kwargs):
...
+        crop_score_indices_name = nms_name + '/Crop_scores_'
+        crop_score_indices = Crop(graph, {'name': crop_score_indices_name, 'axis': int64_array([1]),
+                                          'offset': int64_array([2]), 'dim': int64_array([1])}).create_node()
+        nms.out_port(1).get_connection().insert_node(crop_score_indices)
+        squeeze_output_scores = create_op_node_with_second_input(graph, Squeeze, int64_array([1]),
+                                                                 {'name': crop_score_indices_name + '/Squeeze'})
+        crop_score_indices.out_port(0).get_connection().insert_node(squeeze_output_scores)
+
+        # make inputs 2 to 6 to have shape [1] instead of [0] (convert 0D to 1D)
+        for port_id in range(2, 6):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/non_max_suppression_normalize.py:93]: This is change is already done in the `model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py`

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py (hunk 1 above)
Comment (vgavrilo) [reply model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py]: Fixed.

File: model-optimizer/extensions/front/tf/non_max_suppression_normalize.py (hunk 4 above)
Comment (vgavrilo) [reply model-optimizer/extensions/front/tf/non_max_suppression_normalize.py:93]: The transformation from the file `model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py` performs this change only for ONNX models, but the transformation from the file `model-optimizer/extensions/front/tf/non_max_suppression_normalize.py` performs changes only for TF models.

File: model-optimizer/extensions/ops/non_max_suppression.py (hunk 2 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py]: Fixed.

File: model-optimizer/extensions/ops/non_max_suppression.py (hunk 3 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py:104]: If we write 
```python3
            if not node.out_port(1).disconnected(): 
                node.out_port(1).set_data_type(np.float32)
            if not node.out_port(2).disconnected():
                node.out_port(2).set_data_type(np.int64)
```
instead of lines 104-107, the we can get
```sh
mo.utils.error.Error: The data type for out port 2 of node ... is not defined
```
for a node with `type = `NonMaxSuppression` .

The reason of  that is lines 271-272
```python3
            log.debug('Type infer for node {}: {}'.format(node_name,
                                                          [port.get_data_type() for port in node.out_ports().values()]))
```
in the file `model-optimizer\mo\middle\passes\infer.py`.

File: model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py [hunk 5]
```diff
@@ -30,10 +30,10 @@ class NonMaxSuppressionNormalize(FrontReplacementSubgraph):
 
     def find_and_replace_pattern(self, graph: Graph):
         for nms in graph.get_op_nodes(op='NonMaxSuppression'):
-            # make inputs 2 to 4 to have shape [] instead of [1] (convert 1D to 0D)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py:33]: Update the comment in the transformation description.

File: model-optimizer/extensions/front/tf/non_max_suppression_normalize.py [hunk 6]
```diff
@@ -64,18 +64,34 @@ def replace_sub_graph(graph: Graph, match: dict, **kwargs):
                                                                 {'name': nms.soft_get('name') + '/Unsqueeze_1'})
         nms.in_port(1).get_connection().insert_node(unsqueeze_box_scores)
 
-        # prepare output
-        crop_box_indices = Crop(graph, {'name': nms.soft_get('name') + '/Crop', 'axis': int64_array([1]),
+        nms_name = nms.soft_get('name', nms.id)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/non_max_suppression_normalize.py:67]: Rewrite this transformation to not use the `pattern` function.

File: model-optimizer/extensions/front/tf/non_max_suppression_normalize.py (hunk 4 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/tf/non_max_suppression_normalize.py:93]: Move `model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py` to `model-optimizer/extensions/front/non_max_suppression_normalize.py` so it will be applied for any framework models and remove the duplicate code in this file.

File: model-optimizer/extensions/ops/non_max_suppression.py [hunk 7]
```diff
@@ -47,7 +46,7 @@ def __init__(self, graph: Graph, attrs: dict):
 
     def backend_attrs(self):
         version = self.get_opset()
-        if version in ['opset3', 'opset4']:
+        if version in ['opset3', 'opset4', 'opset5']:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/non_max_suppression.py:49]: Please, make sure that the IR containing NMS-4 can be read and written by MO IR reader correctly.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py (hunk 5 above)
Comment (vgavrilo) [reply model-optimizer/extensions/front/onnx/non_max_suppression_normalize.py:33]: Done.

File: model-optimizer/extensions/front/tf/non_max_suppression_normalize.py (hunk 6 above)
Comment (vgavrilo) [reply model-optimizer/extensions/front/tf/non_max_suppression_normalize.py]: Done.

File: model-optimizer/extensions/front/tf/non_max_suppression_normalize.py (hunk 4 above)
Comment (vgavrilo) [reply model-optimizer/extensions/front/tf/non_max_suppression_normalize.py]: Done.

File: model-optimizer/extensions/ops/non_max_suppression.py (hunk 7 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py:48]: Done.

**Review by pavel-esir — APPROVED** [review]
Looks good to me.

File: model-optimizer/extensions/ops/non_max_suppression.py (hunk 3 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/ops/non_max_suppression.py:103]: You may have 3 output ports but only 2 of them connected: 0 and 2 (1 is not connected). Your check for `num_of_outputs` will be incorrect.

You must check the the output port existence using `has_port` method and only if it exist then update the data type.

File: model-optimizer/extensions/ops/non_max_suppression.py [hunk 8]
```diff
@@ -57,31 +55,56 @@ def backend_attrs(self):
...
         boxes_shape = node.in_port(0).data.get_shape()
         assert boxes_shape is not None, 'The shape of tensor with boxes is not defined'
         scores_shape = node.in_port(1).data.get_shape()
         assert scores_shape is not None, 'The shape of tensor with scores is not defined'
         assert len(boxes_shape) == 3, 'Length of tensors with boxes must be equal to 3'
         assert len(scores_shape) == 3, 'Length of tensors with scores must be equal to 3'
 
-        max_output_boxes_per_class = node.in_port(2).data.get_value()
-        if max_output_boxes_per_class is None:
+        if num_of_inputs >= 3:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/non_max_suppression.py:72]: Please, add a comment describing this check and why the `max_output_boxes_per_class = 0` for the `else` case

File: model-optimizer/extensions/ops/non_max_suppression.py [hunk 9]
```diff
@@ -57,31 +55,56 @@ def backend_attrs(self):
...
-        if node.get_opset() in ['opset3', 'opset4']:
+        opset = node.get_opset()
+        if opset == 'opset5':
+            node.out_port(0).set_data_type(node.output_type)
+            num_of_outputs = len(node.out_ports())
+            if num_of_outputs >= 2:
+                node.out_port(1).set_data_type(np.float32)
+            if num_of_outputs >= 3:
+                node.out_port(2).set_data_type(np.int64)
+        elif node.get_opset() in ['opset3', 'opset4']:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/non_max_suppression.py:107]: replace `node.get_opset()` with `opset` variable

File: model-optimizer/extensions/ops/non_max_suppression.py [hunk 10]
```diff
@@ -57,31 +55,56 @@ def backend_attrs(self):
...
             max_number_of_boxes = min(num_input_boxes, max_output_boxes_per_class) * boxes_shape[0] * num_classes
         else:
             max_number_of_boxes = min(num_input_boxes, boxes_shape[0] * max_output_boxes_per_class * num_classes)
         node.out_port(0).data.set_shape(int64_array([max_number_of_boxes, 3]))
 
+        num_of_outputs = len([port for port in node.out_ports().values() if not port.disconnected()])
+        if num_of_outputs >= 2:
+            node.out_port(1).data.set_shape(int64_array([max_number_of_boxes, 3]))
+        if num_of_outputs >= 3:
+            node.out_port(2).data.set_shape(np.array(1).astype(np.int64))
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/non_max_suppression.py:95]: replace `np.array(1).astype(np.int64)` -> `int64_array(1)`

File: model-optimizer/extensions/ops/non_max_suppression.py [hunk 11]
```diff
@@ -57,31 +55,56 @@ def backend_attrs(self):
...
+        if node.get_opset() in ['opset4', 'opset5']:
             max_number_of_boxes = min(num_input_boxes, max_output_boxes_per_class) * boxes_shape[0] * num_classes
         else:
             max_number_of_boxes = min(num_input_boxes, boxes_shape[0] * max_output_boxes_per_class * num_classes)
         node.out_port(0).data.set_shape(int64_array([max_number_of_boxes, 3]))
 
+        num_of_outputs = len([port for port in node.out_ports().values() if not port.disconnected()])
+        if num_of_outputs >= 2:
+            node.out_port(1).data.set_shape(int64_array([max_number_of_boxes, 3]))
+        if num_of_outputs >= 3:
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/non_max_suppression.py:94]: Check that output port exists. The same for line 92

File: model-optimizer/extensions/ops/non_max_suppression.py (hunk 10 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py]: Fixed.

File: model-optimizer/extensions/ops/non_max_suppression.py (hunk 11 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py:94]: Fixed.

File: model-optimizer/extensions/ops/non_max_suppression.py (hunk 9 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py]: Fixed.

**Review by mvafin — APPROVED** [review]

File: model-optimizer/extensions/ops/non_max_suppression.py (hunk 8 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py:74]: Done.

File: model-optimizer/extensions/ops/non_max_suppression.py [hunk 12]
```diff
@@ -31,13 +31,11 @@ def __init__(self, graph: Graph, attrs: dict):
         mandatory_props = {
             'type': self.op,
             'op': self.op,
-            'version': 'opset4',
+            'version': 'opset5',
             'infer': self.infer,
             'output_type': np.int64,
             'center_point_box': 0,
             'box_encoding': 'corner',
-            'in_ports_count': 5,
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/non_max_suppression.py:39]: Why did you remove these two lines? They are needed to properly initialize the number input/output ports for the operation.

Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py:39]: `in_ports_count` was added again. As for `out_ports_count`, NMS-1, NMS-3, NMS-4 have only 1 output port, but NMS-5 has 3 output ports. Moreover, if `out_ports_count == 3`, then we get an error
```sh
File "...\openvino\model-optimizer\mo\main.py", line 298, in main
    ret_code = driver(argv)
  File "...\openvino\model-optimizer\mo\main.py", line 265, in driver
    ret_res = emit_ir(prepare_ir(argv), argv)
  File "...\openvino\model-optimizer\mo\main.py", line 249, in emit_ir
    meta_info=get_meta_info(argv))
  File "...\openvino\model-optimizer\mo\pipeline\common.py", line 208, in prepare_emit_ir
    type_infer(graph)
  File "...\openvino\model-optimizer\mo\middle\passes\infer.py", line 272, in type_infer
    [port.get_data_type() for port in node.out_ports().values()]))
  File "...\openvino\model-optimizer\mo\middle\passes\infer.py", line 272, in <listcomp>
    [port.get_data_type() for port in node.out_ports().values()]))
  File "...\openvino\model-optimizer\mo\graph\port.py", line 366, in get_data_type
    self.node.name))
mo.utils.error.Error: The data type for out port 2 of node non_max_suppression_with_scores/NonMaxSuppressionV5 is not defined

Process finished with exit code 1
``` 
for the model EfficientDet-D7, and an error
```sh
File "...\openvino\model-optimizer\mo\main.py", line 298, in main
    ret_code = driver(argv)
  File "...\openvino\model-optimizer\mo\main.py", line 265, in driver
    ret_res = emit_ir(prepare_ir(argv), argv)
  File "...\openvino\model-optimizer\mo\main.py", line 249, in emit_ir
    meta_info=get_meta_info(argv))
  File "...\openvino\model-optimizer\mo\pipeline\common.py", line 208, in prepare_emit_ir
    type_infer(graph)
  File "...\openvino\model-optimizer\mo\middle\passes\infer.py", line 272, in type_infer
    [port.get_data_type() for port in node.out_ports().values()]))
  File "...\openvino\model-optimizer\mo\middle\passes\infer.py", line 272, in <listcomp>
    [port.get_data_type() for port in node.out_ports().values()]))
  File "...\openvino\model-optimizer\mo\graph\port.py", line 366, in get_data_type
    self.node.name))
mo.utils.error.Error: The data type for out port 1 of node 1162 is not defined

Process finished with exit code 1
```
for [Mask R CNN model](https://onnxzoo.blob.core.windows.net/models/opset_10/mask_rcnn/mask_rcnn_R_50_FPN_1x.onnx) (this model has NMS-4, not NMS-5).

Finally, if `out_ports_count == 1`, then it is incorrect for NMS-5, because NMS-5 has 3 outputs, and, e.g., 2 outputs of 3 are used in the EfficiendDet-D7 model.

Comment (lazarevevgeny) [reply model-optimizer/extensions/ops/non_max_suppression.py:39]: You can set different number of output port for different versions of the NMS operation in the operation constructor.

Comment (vgavrilo) [reply model-optimizer/extensions/ops/non_max_suppression.py:39]: Then we will have an abovementioned error with the TF model EfficientDet-D7, because this model has NMS-5, but uses only 2 of 3 outputs.

**Review by lazarevevgeny — APPROVED** [review]
